apps/arweave/src/ar_pool.erl

```diff
@@ -0,0 +1,764 @@
+	end.
+
+process_partial_solution_field_size(Solution, Ref) ->
+	#mining_solution{
+		nonce_limiter_output = Output,
+		seed = Seed,
+		next_seed = NextSeed,
+		mining_address = MiningAddress,
+		preimage = Preimage,
+		solution_hash = SolutionH,
+		poa1 = #poa{ chunk = C1 },
+		poa2 = #poa{ chunk = C2 }
+	} = Solution,
+	%% We have less strict deserialization in the pool pipeline to simplify
+	%% the pool proxy implementation. Therefore, we validate the field sizes here
+	%% and return the "rejected_bad_poa" status in case of a failure.
+	case {byte_size(Output), byte_size(Seed), byte_size(NextSeed), byte_size(MiningAddress),
+			byte_size(Preimage), byte_size(SolutionH), byte_size(C1), byte_size(C2)} of
+		{32, 32, 32, 32, 32, 32, L1, L2} when L1 =< ?DATA_CHUNK_SIZE, L2 =< ?DATA_CHUNK_SIZE ->
+			%% The second chunk may be either empty or 256 KiB. ar_poa:validate/1 does
+			%% the proper verification - here we simply protect against payload size abuse.
+			%% We are not strict about the first chunk here to simplify tests.
+			process_partial_solution_poa2(Solution, Ref);
+		_ ->
+			#partial_solution_response{ status = <<"rejected_bad_poa">> }
+	end.
+
+process_partial_solution_poa2(Solution, Ref) ->
+	#mining_solution{
+		recall_byte2 = RecallByte2,
+		poa2 = #poa{ chunk = C, data_path = DP, tx_path = TP }
+	} = Solution,
+	case RecallByte2 of
+		undefined ->
+			case {C, DP, TP} of
+				{<<>>, <<>>, <<>>} ->
+					process_partial_solution_partition_number(Solution, Ref);
+				_ ->
+					#partial_solution_response{ status = <<"rejected_bad_poa">> }
+			end;
+		_ ->
+			process_partial_solution_partition_number(Solution, Ref)
+	end.
+
+process_partial_solution_partition_number(Solution, Ref) ->
+	PartitionNumber = Solution#mining_solution.partition_number,
+	PartitionUpperBound = Solution#mining_solution.partition_upper_bound,
+	Max = max(0, PartitionUpperBound div ?PARTITION_SIZE - 1),
+	case PartitionNumber > Max of
+		false ->
+			process_partial_solution_nonce(Solution, Ref);
+		true ->
+			#partial_solution_response{ status = <<"rejected_bad_poa">> }
+	end.
+
+process_partial_solution_nonce(Solution, Ref) ->
+	Max = max(0, (?RECALL_RANGE_SIZE) div ?DATA_CHUNK_SIZE - 1),
+	case Solution#mining_solution.nonce > Max of
+		false ->
+			process_partial_solution_quick_pow(Solution, Ref);
+		true ->
+			#partial_solution_response{ status = <<"rejected_bad_poa">> }
+	end.
+
+process_partial_solution_quick_pow(Solution, Ref) ->
+	#mining_solution{
+		nonce_limiter_output = NonceLimiterOutput,
+		partition_number = PartitionNumber,
+		seed = Seed,
+		mining_address = MiningAddress,
+		preimage = Preimage,
+		solution_hash = SolutionH
+	} = Solution,
+	H0 = ar_block:compute_h0(NonceLimiterOutput, PartitionNumber, Seed, MiningAddress),
+	case ar_block:compute_solution_h(H0, Preimage) of
+		SolutionH ->
+			process_partial_solution_pow(Solution, Ref, H0);
+		_ ->
+			#partial_solution_response{ status = <<"rejected_wrong_hash">> }
```

Comment:
Will this branch ever be triggered? It looks like `compute_solution_h` just does:
```
crypto:hash(sha256, << H0:32/binary, Preimage/binary >>)
```

Reply:
Added a clarifying comment